### DISCUSSION TITLE
HTML Date field + up-validate

### DISCUSSION THREAD

**Discussion by sturatcliffe** [unpoly/unpoly#334] [General]
Hi all,

Has anyone else noticed some weird behaviour when using ```up-validate``` on a HTML date field? I've only checked in Chrome so far, but when you try to enter a date, it seems to trigger the validation request when entering the first character of the year part.

This means you end up not being able to complete the year part as when the validate request resolves and the form is re-rendered, the focus on the date field resets so you end up in an infinite loop of entering the day + month + first digit of the year.

Hopefully this GIF helps demonstrate it!

![up-validate-date](https://user-images.githubusercontent.com/17248652/146397836-79ad1b68-2596-4d2b-8ffe-c27c3ca61e61.gif)

Has anyone found a way of stopping this behaviour so we can have Unpoly trigger validation on date/datetime fields?

Cheers.

**Comment by sturatcliffe** [comment]
In case anyone else finds this with the same issue, this is what I've currently done to resolve:

```
up.compiler('input[type=date]', function(element) {
    element.addEventListener('blur', function() {
        up.validate(element);
    });
});
```

You may need to narrow down which inputs this targets depending on requirements though.

**Comment by triskweline** [comment]
This seems to be a [standing issue with date inputs](https://stackoverflow.com/questions/40762549/html5-input-type-date-onchange-event).

What a mess. A `change` event should only be fired when the user commits a value. If we wanted to see inputs as they happen, we would listen to the `input` event.

Unpoly 2.6 will allow the `[up-validate]` event to be configurable through an attribute. Although I don't like fixing browser quirks in Unpoly, I'm thinking about changing the default event `blur` for date inputs.

**Reply by sturatcliffe** [reply]
> Although I don't like fixing browser quirks in Unpoly, I'm thinking about changing the default event `blur` for date inputs.

Agree it's not up to Unpoly to fix browser quirks - especially as there is a workaround by using a simple compiler. Also, if you only change date/time/datetime inputs to listen for ```blur``` rather than ```change```, it may be a little jarring and inconsistent when other inputs only trigger validation if the user actually interacts with the field?

In my case, I prefer the validation to occur on ```blur``` for all form fields, so there is immediate feedback if a user ignores a required field. In which case the compiler is required anyway which negates the date field issue entirely.
